keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,438 @@
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access, where cross-family access may be supported depending upon the chosen ingress controller.
+- CNI network plugins: Plugins other than the Bridge, PTP, and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Multiple IPs vs. Dual-Stack: Code changes will be done in a way to facilitate future expansion to more general multiple-IPs-per-pod and multiple-IPs-per-node support. However, this initial release will impose "dual-stack-centric" IP address limits as follows:
+  - Pod addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
+  - Node addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
+  - Service addresses: 1 IP address pod service
+- For simplicity, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. Each IPv4/IPv6 address that is assigned to a backend pod will be treated and displayed as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
+- Kube-DNS is expected to be End-of-Life soon, so dual-stack testing will be performed using coreDNS.
+- External load balancers that rely on Kubernetes services for load balancing functionality will only work with the IP family that matches the IP family of the cluster's service CIDR.
+- Dual-stack support for Kubernetes orchestration tools other than kubeadm (e.g. miniKube, KubeSpray, etc.) are considered outside of the scope of this proposal.
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods and nodes. Here is a summary of the proposal (details follow in subsequent sections):
+
+- Kubernetes needs to be made aware of multiple IPs per pod (limited to one IPv4 and one IPv6 address per pod maximum).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
```

Comment:
I do not think that duplicate pod endpoints is going to be acceptable. It affects more than just load balancing and DNS. Distributed applications use them for peer discovery. Monitoring systems such as Prometheus use them to identify monitoring targets. Users use them to determine how many instances are up.

Reply:
If this is the case, then the [Endpoints V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will need to change, specifically for the Subsets. I need to think through the details, but this would have the same backwards compatibility concerns as podIP/podIPs. The newer, or internal list of Subsets would have to couple each IPv4 address with its partner IPv6 (i.e. a V4/V6 pair from a given pod). This would then require all consumers (clients) of the Endpoints API to eventually use the dual-stack information... that's going to take some time.

Reply:
+1 for @bgrant0607's comments.

Could we change endpoints API to support multiple addresses for each subset?

Reply:
Yes, I'll rewrite this to add support for multiple addresses per subset.